File: src/api/auth/token_engine.py
import base64
import os
import secrets
import time
import uuid
from typing import Any, Dict, Optional

import jwt
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ed25519

from src.api.errors import AxiomException, ErrorCodes


class TokenEngine:
    """Ed25519 JWT and Refresh Token generator/validator."""

    def __init__(self, keys_dir: str = "./data/auth/.keys"):
        self.keys_dir = keys_dir
        self.private_key = None
        self.public_key = None
        self.key_id = "axiom-auth-1"

    def init_keys(self) -> None:
        """Loads or generates the Ed25519 key pair."""
        os.makedirs(self.keys_dir, exist_ok=True)
        key_path = os.path.join(self.keys_dir, "ed25519.pem")

        if os.path.exists(key_path):
            with open(key_path, "rb") as f:
                key_data = f.read()
                key = serialization.load_pem_private_key(
                    key_data, password=None, backend=default_backend()
                )
            if not isinstance(key, ed25519.Ed25519PrivateKey):
                raise ValueError("Invalid key type loaded")
            self.private_key = key
        else:
            self.private_key = ed25519.Ed25519PrivateKey.generate()
            pem = self.private_key.private_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PrivateFormat.PKCS8,
                encryption_algorithm=serialization.NoEncryption(),
            )
            with open(key_path, "wb") as f:
                f.write(pem)

        self.public_key = self.private_key.public_key()

    def create_access_token(
        self,
        project_id: str,
        uid: str,
        email: str,
        email_verified: bool,
        is_anonymous: bool,
        totp_verified: bool,
        ttl: int,
        custom_claims: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Generates an Ed25519 signed JWT."""
        if not self.private_key:
            self.init_keys()
        assert self.private_key is not None

        now = int(time.time())
        payload = {
            "sub": uid,
            "email": email,
            "email_verified": email_verified,
            "is_anonymous": is_anonymous,
            "totp_verified": totp_verified,
            "iat": now,
            "exp": now + ttl,
            "iss": "axiom",
            "aud": project_id,
            "jti": str(uuid.uuid4()),
        }

        if custom_claims:
            for k, v in custom_claims.items():
                if k not in payload:
                    payload[k] = v

        return jwt.encode(
            payload,
            self.private_key,
            algorithm="EdDSA",
            headers={"kid": self.key_id},
        )

    def verify_access_token(self, token: str, project_id: str) -> Dict[str, Any]:
        """Verifies JWT signature, expiry, and audience (project isolation)."""
        if not self.public_key:
            self.init_keys()
        assert self.public_key is not None

        try:
            decoded = jwt.decode(
                token,
                self.public_key,
                algorithms=["EdDSA"],
                audience=project_id,
                issuer="axiom",
            )
            return decoded
        except jwt.ExpiredSignatureError:
            raise AxiomException(
                code=ErrorCodes.AUTH_TOKEN_EXPIRED,
                message="Access token expired",
                status_code=401,
            )
        except jwt.InvalidAudienceError:
            raise AxiomException(
                code=ErrorCodes.AUTH_TOKEN_INVALID,
                message="Invalid audience for this project",
                status_code=401,
            )
        except jwt.PyJWTError as e:
            raise AxiomException(
                code=ErrorCodes.AUTH_TOKEN_INVALID,
                message=f"Invalid token: {str(e)}",
                status_code=401,
            )

    def get_jwks(self) -> Dict[str, Any]:
        """Returns the public key in JWKS format for third-party verification."""
        if not self.public_key:
            self.init_keys()
        assert self.public_key is not None

        # Extract raw public bytes
        pub_bytes = self.public_key.public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )

        # Base64url encode without padding
        x = base64.urlsafe_b64encode(pub_bytes).rstrip(b"=").decode("ascii")

        return {
            "keys": [
                {
                    "kty": "OKP",
                    "crv": "Ed25519",
                    "x": x,
                    "kid": self.key_id,
                    "use": "sig",
                }
            ]
        }

    @staticmethod
    def generate_refresh_token() -> str:
        """Returns a 64-byte random hex string."""
        return secrets.token_hex(64)


# Global singleton
token_engine = TokenEngine()

File: src/api/auth/token_engine.rs
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use jsonwebtoken::{decode, encode, Algorithm, DecodingKey, EncodingKey, Header, Validation};
use once_cell::sync::Lazy;
use rand::Rng;
use ring::rand::SystemRandom;
use ring::signature::Ed25519KeyPair;
use ring::signature::KeyPair as RingKeyPair;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};
use tokio::sync::RwLock;
use tracing::info;

const KEY_ID: &str = "axiom-auth-1";
const KEYS_DIR: &str = "data/auth/.keys";

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct JwtClaims {
    pub sub: String,
    pub email: String,
    pub email_verified: bool,
    pub is_anonymous: bool,
    pub totp_verified: bool,
    pub iat: i64,
    pub exp: i64,
    pub iss: String,
    pub aud: String,
    pub jti: String,
    #[serde(flatten)]
    pub extra: HashMap<String, serde_json::Value>,
}

struct KeyPair {
    encoding_key: EncodingKey,
    decoding_key: DecodingKey,
    public_bytes: Vec<u8>,
}

static KEY_PAIR: Lazy<RwLock<Option<KeyPair>>> = Lazy::new(|| RwLock::new(None));

fn now_secs() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs() as i64
}

pub async fn init_keys() {
    let mut lock = KEY_PAIR.write().await;
    if lock.is_some() {
        return;
    }

    fs::create_dir_all(KEYS_DIR).expect("Failed to create auth key directory");
    let key_path = Path::new(KEYS_DIR).join("ed25519.pk8");

    let pkcs8_bytes = if key_path.exists() {
        fs::read(&key_path).expect("Failed to read existing key file")
    } else {
        let rng = SystemRandom::new();
        let pkcs8 = Ed25519KeyPair::generate_pkcs8(&rng).expect("Failed to generate pkcs8");
        let bytes = pkcs8.as_ref().to_vec();
        fs::write(&key_path, &bytes).expect("Failed to write key file");
        info!("Generated new Ed25519 signing key at {:?}", key_path);
        bytes
    };

    let key_pair = Ed25519KeyPair::from_pkcs8(&pkcs8_bytes).expect("Invalid PKCS8 bytes");
    let pub_bytes = key_pair.public_key().as_ref().to_vec();

    let encoding_key = EncodingKey::from_ed_der(&pkcs8_bytes);
    let decoding_key = DecodingKey::from_ed_der(&pub_bytes);

    *lock = Some(KeyPair {
        encoding_key,
        decoding_key,
        public_bytes: pub_bytes,
    });
}

pub async fn create_access_token(
    project_id: &str,
    uid: &str,
    email: &str,
    email_verified: bool,
    is_anonymous: bool,
    totp_verified: bool,
    ttl_secs: i64,
    extra: HashMap<String, serde_json::Value>,
) -> Result<String, String> {
    let lock = KEY_PAIR.read().await;
    let kp = lock.as_ref().ok_or("Auth keys not initialized")?;

    let claims = JwtClaims {
        sub: uid.to_string(),
        email: email.to_string(),
        email_verified,
        is_anonymous,
        totp_verified,
        iat: now_secs(),
        exp: now_secs() + ttl_secs,
        iss: "axiom".to_string(),
        aud: project_id.to_string(),
        jti: uuid::Uuid::new_v4().to_string(),
        extra,
    };

    let mut header = Header::new(Algorithm::EdDSA);
    header.kid = Some(KEY_ID.to_string());

    encode(&header, &claims, &kp.encoding_key).map_err(|e| e.to_string())
}

pub async fn verify_access_token(token: &str, project_id: &str) -> Result<JwtClaims, String> {
    let lock = KEY_PAIR.read().await;
    let kp = lock.as_ref().ok_or("Auth keys not initialized")?;

    let mut validation = Validation::new(Algorithm::EdDSA);
    validation.set_audience(&[project_id]);
    validation.set_issuer(&["axiom"]);

    decode::<JwtClaims>(token, &kp.decoding_key, &validation)
        .map(|td| td.claims)
        .map_err(|e| e.to_string())
}

pub async fn get_jwks() -> serde_json::Value {
    let lock = KEY_PAIR.read().await;
    let pub_bytes = lock
        .as_ref()
        .map(|kp| kp.public_bytes.clone())
        .unwrap_or_default();
    let x = URL_SAFE_NO_PAD.encode(&pub_bytes);

    serde_json::json!({
        "keys": [{
            "kty": "OKP",
            "crv": "Ed25519",
            "x": x,
            "kid": KEY_ID,
            "use": "sig",
            "alg": "EdDSA"
        }]
    })
}

pub fn generate_refresh_token() -> String {
    use rand::distributions::Alphanumeric;
    let token: String = rand::thread_rng()
        .sample_iter(&Alphanumeric)
        .take(64)
        .map(char::from)
        .collect();
    token
}

pub fn sha256_hex(data: &str) -> String {
    use sha2::{Digest, Sha256};
    let hash = Sha256::digest(data.as_bytes());
    hex::encode(hash)
}
